torch/csrc/jit/passes/python_print.cpp

```diff
@@ -167,7 +167,74 @@ const static std::unordered_set<std::string> reserved_names = {
 };
 
 struct PythonPrintPass {
-  std::ostringstream body_;
+  using SourceRangeStack = std::vector<SourceRange>;
+  SourceRangeStack source_range_stack_ = {SourceRange("")};
+
+  struct WithSourceRange {
+    explicit WithSourceRange(SourceRangeStack* stack, SourceRange sr)
+        : stack(stack) {
+      TORCH_INTERNAL_ASSERT(stack);
+      stack->push_back(std::move(sr));
+    }
+
+    ~WithSourceRange() {
+      stack->pop_back();
+    }
+
+    SourceRangeStack* stack;
+  };
+
+  class TaggedStringStream {
+   public:
+    TaggedStringStream(const SourceRangeStack* srs) : srs(srs) {}
+    TaggedStringStream(TaggedStringStream&& rhs) = default;
+
+    TaggedStringStream& operator<<(const std::string& s) {
+      ranges_.emplace_back((size_t)oss.tellp(), srs->back());
+      oss << s;
+      return *this;
+    }
+
+    TaggedStringStream& operator<<(const TaggedStringStream& rhs) {
+      for (const auto& range : rhs.ranges_) {
+        ranges_.emplace_back((size_t)oss.tellp() + range.bytes, range.range);
+      }
+      oss << rhs.oss.str();
+      return *this;
+    }
+
+    // This overload is here to prevent people from shooting themselves in the
+    // foot. I would be highly surprised if someone actually wanted to write out
+    // the address of a TaggedStringStream in the pretty print.
+    TaggedStringStream& operator<<(
+        const std::shared_ptr<TaggedStringStream>& rhs) {
+      (*this) << *rhs;
+      return *this;
+    }
+
+    template <typename T>
+    TaggedStringStream& operator<<(const T& t) {
```

Comment:
This is also inefficient. It is allocating a new buffer _every time_ << is called. Can we just record it directly into the stream? Even in the real operator it just writes to the real stream anyway.